snuba/web/query.py

```diff
@@ -133,26 +135,87 @@ def _run_query_pipeline(
           this comment, that may no longer be true
 
     """
-    if request.query_settings.get_dry_run():
-        query_runner = _dry_run_query_runner
-    else:
-        query_runner = partial(
-            _run_and_apply_column_names,
-            timer=timer,
-            query_metadata=query_metadata,
-            attribution_info=request.attribution_info,
-            robust=robust,
-            concurrent_queries_gauge=concurrent_queries_gauge,
-        )
 
     record_missing_use_case_id(request, dataset)
     record_subscription_created_missing_tenant_ids(request)
 
-    return (
-        dataset.get_query_pipeline_builder()
-        .build_execution_pipeline(request, query_runner)
-        .execute()
+    from snuba.datasets.entities.factory import get_entity
+    from snuba.datasets.plans.query_plan import QueryRunner
+    from snuba.pipeline.composite import CompositeExecutionPipeline
+    from snuba.pipeline.query_pipeline import QueryPipelineStage
+
+    class EntityAndStoragePipelineStage(
+        QueryPipelineStage[Request, tuple[Query, QuerySettings, Reader, str]]
+    ):
+        def _execute(
+            self, input: QueryPipelineResult[Request]
+        ) -> QueryPipelineResult[tuple[Query, QuerySettings, Reader, str]]:
+            _clickhouse_query = None
+            _reader = None
+            _cluster_name = None
+            _query_settings = None
+
+            def _query_runner(
+                clickhouse_query: Union[Query, CompositeQuery[Table]],
+                query_settings: QuerySettings,
+                reader: Reader,
+                cluster_name: str,
+            ) -> QueryResult:
+                nonlocal _clickhouse_query, _reader, _cluster_name, _query_settings
+                _clickhouse_query = clickhouse_query
+                _reader = reader
+                _query_settings = query_settings
+                _cluster_name = cluster_name
+
+            if isinstance(input.data.query, LogicalQuery):
+                entity = get_entity(input.data.query.get_from_clause().key)
+                execution_pipeline = (
+                    entity.get_query_pipeline_builder().build_execution_pipeline(
+                        request, _query_runner
+                    )
+                )
+            else:
+                execution_pipeline = CompositeExecutionPipeline(
+                    input.data.query, input.data.query_settings, _query_runner
+                )
+
+            execution_pipeline.execute()
```

Comment:
If I'm not mistaken, calling `execute()` on the execution pipeline will also call the query runner passed into the pipeline.

Reply:
yes but we passed in a query runner that only sets a variable. It does not run the query! This is little bit of magic that allows us to actually decouple execution from processing. This was not possible before because query splitters were a thing and used the runner multiple times